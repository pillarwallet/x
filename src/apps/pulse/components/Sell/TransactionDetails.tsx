import { format } from 'date-fns';
import { useState } from 'react';
import ProgressStep from './ProgressStep';
import TransactionErrorBox from './TransactionErrorBox';
import TransactionInfo from './TransactionInfo';

interface TransactionDetailsProps {
  onDone: () => void;
  userOpHash: string;
  chainId: number;
  status:
    | 'Starting Transaction'
    | 'Transaction Pending'
    | 'Transaction Complete'
    | 'Transaction Failed';
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: {
    symbol: string;
    name: string;
    logo: string;
  } | null;
  tokenAmount?: string;
  sellOffer?: {
    tokenAmountToReceive: number;
    minimumReceive: number;
  } | null;
  // Timestamps for when each step completed
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  // Buy-specific timestamps
  resourceLockCompletedAt?: Date;
  // Transaction details from getUserOperationStatus
  txHash?: string;
  gasFee?: string;
  errorDetails?: string;
  // Buy-specific hashes
  resourceLockTxHash?: string;
  completedTxHash?: string;
  resourceLockChainId?: number;
  completedChainId?: number;
  // Failure indicators
  isResourceLockFailed?: boolean;
}

const TransactionDetails = ({
  onDone,
  userOpHash,
  chainId,
  status,
  isBuy = false,
  sellToken,
  tokenAmount,
  sellOffer,
  submittedAt,
  pendingCompletedAt,
  resourceLockCompletedAt,
  txHash,
  gasFee,
  errorDetails,
  resourceLockTxHash,
  completedTxHash,
  resourceLockChainId,
  completedChainId,
  isResourceLockFailed = false,
}: TransactionDetailsProps) => {
  const [showTooltip, setShowTooltip] = useState<boolean>(false);

  // Function to determine step status based on current transaction status
  const getStepStatus = (
    step: 'Submitted' | 'Pending' | 'ResourceLock' | 'Completed'
  ) => {
    if (status === 'Starting Transaction') {
      // All steps pending
      return 'pending';
    }

    if (status === 'Transaction Pending') {
      if (step === 'Submitted') return 'completed';
      if (step === 'Pending') {
        return 'pending';
      }
      if (step === 'ResourceLock') {
        // For Buy, resource lock step shows pending until we have the hash
        if (isBuy) {
          return resourceLockTxHash ? 'completed' : 'pending';
        }
        return 'pending';
      }
      if (step === 'Completed') {
        // For Buy, completed step shows pending only if resource lock is done
        if (isBuy) {
          if (!resourceLockTxHash) {
            // If resource lock is not done yet, completed step should be inactive
            return 'inactive';
          }
          return completedTxHash ? 'completed' : 'pending';
        }
        // For Sell, completed step should be inactive when pending is active
        return 'inactive';
      }
    }

    if (status === 'Transaction Complete') {
      if (step === 'Submitted') return 'completed';
      if (step === 'Pending') return 'completed';
      if (step === 'ResourceLock') return 'completed';
      if (step === 'Completed') return 'completed';
    }

    if (status === 'Transaction Failed') {
      if (step === 'Submitted') return 'completed';
      if (step === 'Pending') {
        return 'completed';
      }
      if (step === 'ResourceLock') {
        if (isBuy && isResourceLockFailed) return 'failed';
        return 'completed';
      }
      if (step === 'Completed') {
        if (isBuy && isResourceLockFailed) return 'inactive';
        return 'failed'; // For Sell, Completed should be failed when transaction fails
      }
    }

    return 'pending';
  };

  // Function to format timestamps for completed steps
  const getStepTimestamp = (
    step: 'Submitted' | 'Pending' | 'ResourceLock' | 'Completed'
  ) => {
    const stepStatus = getStepStatus(step);

    // Only show timestamps for completed steps, not the last step
    if (stepStatus !== 'completed' || step === 'Completed') return undefined;

    if (step === 'Submitted' && submittedAt) {
      // For submitted step, show full date and time
      return `${format(submittedAt, 'MMM d, yyyy')} â€¢ ${format(
        submittedAt,
        'HH:mm'
      )}`;
    }

    if (step === 'Pending' && pendingCompletedAt) {
      // For Sell pending step, show pendingCompletedAt timestamp
      return format(pendingCompletedAt, 'HH:mm');
    }

    if (step === 'ResourceLock') {
      // For Buy show resource lock timestamp if present, else pendingCompletedAt fallback
      const ts = isBuy
        ? resourceLockCompletedAt || pendingCompletedAt
        : pendingCompletedAt;
      if (ts) return format(ts, 'HH:mm');
    }

    return undefined;
  };

  return (
    <div className="flex flex-col gap-4 h-full w-full">
      <p className="text-xl text-white font-normal">Transaction Details</p>
      <div className="flex flex-col h-full w-full rounded-[10px] border border-dashed border-[#25232D] border-[4px_4px] p-6 items-center justify-center">
        <p className="text-5xl text-white font-medium">
          {isBuy ? (
            <span className="text-white/30">$</span>
          ) : (
            <div className="relative inline-block">
              <span
                className="text-white/30 cursor-help"
                onMouseEnter={() => setShowTooltip(true)}
                onMouseLeave={() => setShowTooltip(false)}
              >
                {(() => {
                  const symbol = sellToken?.symbol || 'Token';
                  return symbol.length > 3
                    ? `${symbol.slice(0, 3)}...`
                    : symbol;
                })()}
              </span>
              {showTooltip &&
                sellToken?.symbol &&
                sellToken.symbol.length > 3 && (
                  <div className="absolute bottom-full left-1/2 transform -translate-x-1/2 mb-2 px-2 py-1 bg-black text-white text-xs rounded shadow-lg z-10 whitespace-nowrap">
                    {sellToken.symbol}
                  </div>
                )}
            </div>
          )}{' '}
          {tokenAmount || '0'}
        </p>
        <div className="flex gap-1.5">
          <img
            className="w-4 h-4 rounded"
            src="/src/apps/pulse/assets/usd-coin-usdc-logo.png"
            alt="USDC"
          />
          <p className="text-white font-normal text-[13px]">
            {sellOffer?.tokenAmountToReceive?.toFixed(6) || '0'}{' '}
            <span className="text-white/30">USDC</span>
          </p>
        </div>
      </div>

      {status !== 'Transaction Complete' && (
        <div className="flex h-full w-full rounded-[10px] border border-dashed border-[#25232D] border-[4px_4px] p-3">
          <div className="flex flex-col items-center justify-center w-full">
            {/* Progress Bar */}
            <div className="flex flex-col w-full">
              <ProgressStep
                step="Submitted"
                status={getStepStatus('Submitted')}
                label="Submitted"
                lineStatus={
                  isBuy
                    ? getStepStatus('ResourceLock')
                    : getStepStatus('Pending')
                }
                timestamp={getStepTimestamp('Submitted')}
              />

              {isBuy ? (
                <>
                  <ProgressStep
                    step="ResourceLock"
                    status={getStepStatus('ResourceLock')}
                    label="Resource Lock"
                    lineStatus={getStepStatus('Completed')}
                    timestamp={getStepTimestamp('ResourceLock')}
                  />

                  <ProgressStep
                    step="Completed"
                    status={getStepStatus('Completed')}
                    label={
                      getStepStatus('Completed') === 'failed'
                        ? 'Transaction Failed'
                        : 'Completed'
                    }
                    isLast
                    timestamp={getStepTimestamp('Completed')}
                  />
                </>
              ) : (
                <>
                  <ProgressStep
                    step="Pending"
                    status={getStepStatus('Pending')}
                    label="Pending"
                    lineStatus={getStepStatus('Completed')}
                    timestamp={getStepTimestamp('Pending')}
                  />

                  <ProgressStep
                    step="Completed"
                    status={getStepStatus('Completed')}
                    label={
                      getStepStatus('Completed') === 'failed'
                        ? 'Transaction Failed'
                        : 'Completed'
                    }
                    isLast
                    timestamp={getStepTimestamp('Completed')}
                  />
                </>
              )}
            </div>
          </div>
        </div>
      )}

      <div className="flex flex-col h-full w-full rounded-[10px] border border-dashed border-[#25232D] border-[4px_4px] p-3">
        <TransactionInfo
          status={status}
          userOpHash={userOpHash}
          txHash={txHash}
          chainId={chainId}
          gasFee={gasFee}
          completedAt={pendingCompletedAt}
          isBuy={isBuy}
          resourceLockTxHash={resourceLockTxHash}
          completedTxHash={completedTxHash}
          resourceLockChainId={resourceLockChainId}
          completedChainId={completedChainId}
        />
      </div>

      {status === 'Transaction Failed' && (
        <TransactionErrorBox technicalDetails={errorDetails} />
      )}

      <button
        className="px-6 py-2 bg-[#8A77FF] text-white rounded-lg font-normal text-[14px]"
        type="button"
        onClick={onDone}
      >
        Done
      </button>
    </div>
  );
};

export default TransactionDetails;
