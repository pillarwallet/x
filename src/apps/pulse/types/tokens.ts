type ParsedToken = {
  name: string;
  symbol: string;
  markrtCap: number;
  liquidity: number;
  volume: number;
  logo: string;
  price: number;
  priceChange: number;
  blockchain: string;
}

type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
}