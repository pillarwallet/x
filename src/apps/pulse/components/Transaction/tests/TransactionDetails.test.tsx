/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// types
import { SelectedToken } from '../../../types/tokens';

// components
import TransactionDetails from '../TransactionDetails';

// Mock dependencies
vi.mock('../../../../hooks/useTransactionKit', () => ({
  default: () => ({
    walletAddress: '0x1234567890123456789012345678901234567890',
  }),
}));

vi.mock('../../../../utils/number', () => ({
  formatExponentialSmallNumber: (num: number) => num.toFixed(2),
  limitDigitsNumber: (num: number) => num,
}));

vi.mock('../ProgressStep', () => ({
  default: ({ step, status, label, timestamp }: any) => (
    <div data-testid={`progress-step-${step}`}>
      <span data-testid={`step-${step}-status`}>{status}</span>
      <span data-testid={`step-${step}-label`}>{label}</span>
      {timestamp && (
        <div data-testid={`step-${step}-timestamp`}>{timestamp}</div>
      )}
    </div>
  ),
}));

vi.mock('../TransactionErrorBox', () => ({
  default: ({ technicalDetails }: { technicalDetails: string }) => (
    <div data-testid="transaction-error-box">
      <div data-testid="technical-details">{technicalDetails}</div>
    </div>
  ),
}));

vi.mock('../TransactionInfo', () => ({
  default: ({ status, userOpHash, txHash, chainId, gasFee }: any) => (
    <div data-testid="transaction-info">
      <span data-testid="info-status">{status}</span>
      <span data-testid="info-userOpHash">{userOpHash}</span>
      <span data-testid="info-txHash">{txHash}</span>
      <span data-testid="info-chainId">{chainId}</span>
      <span data-testid="info-gasFee">{gasFee}</span>
    </div>
  ),
}));

vi.mock('../Misc/CloseButton', () => ({
  default: ({ onClose }: { onClose: () => void }) => (
    <button type="button" onClick={onClose} data-testid="close-button">
      Close
    </button>
  ),
}));

vi.mock('../Misc/Esc', () => ({
  default: ({ onClose }: { onClose: () => void }) => (
    <button type="button" onClick={onClose} data-testid="esc-button">
      ESC
    </button>
  ),
}));

const mockSellToken: SelectedToken = {
  name: 'Test Token',
  symbol: 'TEST',
  logo: 'test-logo.png',
  usdValue: '100.00',
  dailyPriceChange: 0.05,
  chainId: 1,
  decimals: 18,
  address: '0x1234567890123456789012345678901234567890',
};

const mockBuyToken: SelectedToken = {
  name: 'Buy Token',
  symbol: 'BUY',
  logo: 'buy-logo.png',
  usdValue: '50.00',
  dailyPriceChange: 0.02,
  chainId: 1,
  decimals: 18,
  address: '0x9876543210987654321098765432109876543210',
};

const mockPayingTokens = [
  {
    totalUsd: 100.0,
    name: 'USD Coin',
    symbol: 'USDC',
    logo: 'usdc-logo.png',
    actualBal: '100.00',
    totalRaw: '100000000',
    chainId: 1,
    address: '0xA0b86a33E6441b8C4C8C0C4C8C0C4C8C0C4C8C0C4',
  },
];

const mockSellOffer = {
  tokenAmountToReceive: 50.0,
  minimumReceive: 45.0,
};

const baseProps = {
  onDone: vi.fn(),
  userOpHash:
    '0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef',
  chainId: 1,
  status: 'Starting Transaction' as const,
  isBuy: false,
  sellToken: mockSellToken,
  buyToken: null,
  tokenAmount: '100',
  sellOffer: mockSellOffer,
  payingTokens: undefined,
  usdAmount: undefined,
  submittedAt: new Date('2023-01-01T00:00:00Z'),
  pendingCompletedAt: undefined,
  resourceLockCompletedAt: undefined,
  txHash: undefined,
  gasFee: '0.001',
  errorDetails: '',
  resourceLockTxHash: undefined,
  completedTxHash: undefined,
  resourceLockChainId: undefined,
  completedChainId: undefined,
  isResourceLockFailed: false,
};

describe('<TransactionDetails />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(<TransactionDetails {...baseProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders different transaction statuses', () => {
    it('displays starting transaction status correctly', () => {
      render(
        <TransactionDetails {...baseProps} status="Starting Transaction" />
      );

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
      expect(screen.getByText('Done')).toBeInTheDocument();
    });

    it('displays transaction pending status correctly', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Pending" />
      );

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
      expect(screen.getByTestId('esc-button')).toBeInTheDocument();
    });

    it('displays transaction complete status correctly', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Complete" />
      );

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
      expect(screen.getByTestId('close-button')).toBeInTheDocument();
    });

    it('displays transaction failed status correctly', () => {
      render(<TransactionDetails {...baseProps} status="Transaction Failed" />);

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
      expect(screen.getByTestId('transaction-error-box')).toBeInTheDocument();
    });
  });

  describe('handles user interactions', () => {
    it('calls onDone when done button is clicked', () => {
      render(<TransactionDetails {...baseProps} />);

      const doneButton = screen.getByText('Done');
      fireEvent.click(doneButton);

      expect(baseProps.onDone).toHaveBeenCalled();
    });

    it('calls onDone when close button is clicked', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Complete" />
      );

      const closeButton = screen.getByTestId('close-button');
      fireEvent.click(closeButton);

      expect(baseProps.onDone).toHaveBeenCalled();
    });

    it('calls onDone when ESC button is clicked', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Pending" />
      );

      const escButton = screen.getByTestId('esc-button');
      fireEvent.click(escButton);

      expect(baseProps.onDone).toHaveBeenCalled();
    });

    it('calls onDone when ESC key is pressed', () => {
      render(<TransactionDetails {...baseProps} />);

      fireEvent.keyDown(document, { key: 'Escape' });

      expect(baseProps.onDone).toHaveBeenCalled();
    });

    it('calls onDone when clicking outside the modal for completed status', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Complete" />
      );

      // Click outside the modal
      fireEvent.mouseDown(document.body);

      expect(baseProps.onDone).toHaveBeenCalled();
    });

    it('does not call onDone when clicking outside the modal for pending status', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Pending" />
      );

      // Click outside the modal
      fireEvent.mouseDown(document.body);

      expect(baseProps.onDone).not.toHaveBeenCalled();
    });
  });

  describe('handles buy mode', () => {
    it('renders buy mode correctly', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
      };

      render(<TransactionDetails {...buyProps} />);

      expect(screen.getByText('50')).toBeInTheDocument();
      expect(screen.getByText('100')).toBeInTheDocument();
      expect(screen.getByText('BUY')).toBeInTheDocument();
    });

    it('shows correct progress steps for buy mode', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
      };

      render(<TransactionDetails {...buyProps} />);

      expect(screen.getByTestId('progress-step-Submitted')).toBeInTheDocument();
      expect(
        screen.getByTestId('progress-step-ResourceLock')
      ).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Completed')).toBeInTheDocument();
    });

    it('passes correct props to TransactionInfo for buy mode', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
        resourceLockTxHash: '0xresource123',
        completedTxHash: '0xcompleted123',
        resourceLockChainId: 1,
        completedChainId: 1,
      };

      render(<TransactionDetails {...buyProps} />);

      expect(screen.getByTestId('transaction-info')).toBeInTheDocument();
    });
  });

  describe('handles sell mode', () => {
    it('renders sell mode correctly', () => {
      render(<TransactionDetails {...baseProps} />);

      expect(screen.getByText('50')).toBeInTheDocument();
      expect(screen.getByText('100')).toBeInTheDocument();
      expect(screen.getByText('TEST')).toBeInTheDocument();
    });

    it('shows correct progress steps for sell mode', () => {
      render(<TransactionDetails {...baseProps} />);

      expect(screen.getByTestId('progress-step-Submitted')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Pending')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Completed')).toBeInTheDocument();
    });

    it('passes correct props to TransactionInfo for sell mode', () => {
      render(<TransactionDetails {...baseProps} />);

      expect(screen.getByTestId('transaction-info')).toBeInTheDocument();
    });
  });

  describe('handles different step statuses', () => {
    it('shows correct step statuses for starting transaction', () => {
      render(
        <TransactionDetails {...baseProps} status="Starting Transaction" />
      );

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'pending'
      );
      expect(screen.getByTestId('step-Pending-status')).toHaveTextContent(
        'pending'
      );
    });

    it('shows correct step statuses for transaction pending', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Pending" />
      );

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-Pending-status')).toHaveTextContent(
        'pending'
      );
    });

    it('shows correct step statuses for transaction complete', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Complete" />
      );

      // Progress steps are hidden when transaction is complete
      expect(
        screen.queryByTestId('step-Submitted-status')
      ).not.toBeInTheDocument();
      expect(
        screen.queryByTestId('step-Pending-status')
      ).not.toBeInTheDocument();
    });

    it('shows correct step statuses for transaction failed', () => {
      render(<TransactionDetails {...baseProps} status="Transaction Failed" />);

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-Pending-status')).toHaveTextContent(
        'completed'
      );
    });
  });

  describe('handles buy mode step statuses', () => {
    it('shows correct step statuses for buy mode starting transaction', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
      };

      render(
        <TransactionDetails {...buyProps} status="Starting Transaction" />
      );

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'pending'
      );
      expect(screen.getByTestId('step-ResourceLock-status')).toHaveTextContent(
        'pending'
      );
      expect(screen.getByTestId('step-Completed-status')).toHaveTextContent(
        'pending'
      );
    });

    it('shows correct step statuses for buy mode transaction pending', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
      };

      render(<TransactionDetails {...buyProps} status="Transaction Pending" />);

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-ResourceLock-status')).toHaveTextContent(
        'pending'
      );
      expect(screen.getByTestId('step-Completed-status')).toHaveTextContent(
        'inactive'
      );
    });

    it('shows correct step statuses for buy mode with resource lock hash', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
        resourceLockTxHash: '0xresource123',
      };

      render(<TransactionDetails {...buyProps} status="Transaction Pending" />);

      expect(screen.getByTestId('step-Submitted-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-ResourceLock-status')).toHaveTextContent(
        'completed'
      );
      expect(screen.getByTestId('step-Completed-status')).toHaveTextContent(
        'pending'
      );
    });
  });

  describe('handles timestamps', () => {
    it('displays submitted timestamp correctly', () => {
      const propsWithSubmitted = {
        ...baseProps,
        status: 'Transaction Pending' as const,
      };

      render(<TransactionDetails {...propsWithSubmitted} />);

      expect(
        screen.getByTestId('step-Submitted-timestamp')
      ).toBeInTheDocument();
    });

    it('displays pending completed timestamp correctly', () => {
      const propsWithPendingCompleted = {
        ...baseProps,
        status: 'Transaction Complete' as const,
        pendingCompletedAt: new Date('2023-01-01T01:00:00Z'),
      };

      render(<TransactionDetails {...propsWithPendingCompleted} />);

      // Progress steps are hidden when transaction is complete
      expect(
        screen.queryByTestId('step-Pending-timestamp')
      ).not.toBeInTheDocument();
    });

    it('displays resource lock completed timestamp correctly for buy mode', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
        status: 'Transaction Complete' as const,
        resourceLockCompletedAt: new Date('2023-01-01T02:00:00Z'),
      };

      render(<TransactionDetails {...buyProps} />);

      // Progress steps are hidden when transaction is complete
      expect(
        screen.queryByTestId('step-ResourceLock-timestamp')
      ).not.toBeInTheDocument();
    });
  });

  describe('handles error states', () => {
    it('displays error box when transaction failed', () => {
      const propsWithError = {
        ...baseProps,
        status: 'Transaction Failed' as const,
        errorDetails: 'Transaction failed due to insufficient gas',
      };

      render(<TransactionDetails {...propsWithError} />);

      expect(screen.getByTestId('transaction-error-box')).toBeInTheDocument();
      expect(screen.getByTestId('technical-details')).toBeInTheDocument();
    });

    it('handles resource lock failure for buy mode', () => {
      const buyPropsWithFailure = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
        isResourceLockFailed: true,
        status: 'Transaction Failed' as const,
      };

      render(<TransactionDetails {...buyPropsWithFailure} />);

      expect(screen.getByTestId('step-ResourceLock-status')).toHaveTextContent(
        'failed'
      );
      expect(screen.getByTestId('step-Completed-status')).toHaveTextContent(
        'inactive'
      );
    });
  });

  describe('handles edge cases', () => {
    it('handles missing optional props gracefully', () => {
      const minimalProps = {
        onDone: vi.fn(),
        userOpHash: '0x1234567890abcdef',
        chainId: 1,
        status: 'Starting Transaction' as const,
      };

      render(<TransactionDetails {...minimalProps} />);

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
    });

    it('handles null token values', () => {
      const propsWithNullTokens = {
        ...baseProps,
        sellToken: null,
        buyToken: null,
      };

      render(<TransactionDetails {...propsWithNullTokens} />);

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
    });

    it('handles missing sell offer', () => {
      const propsWithoutSellOffer = {
        ...baseProps,
        sellOffer: null,
      };

      render(<TransactionDetails {...propsWithoutSellOffer} />);

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
    });

    it('handles missing token amount', () => {
      const propsWithoutTokenAmount = {
        ...baseProps,
        tokenAmount: undefined,
      };

      render(<TransactionDetails {...propsWithoutTokenAmount} />);

      expect(screen.getByText('0')).toBeInTheDocument();
    });
  });

  describe('handles different chain IDs', () => {
    it('renders with different chain ID', () => {
      const propsWithDifferentChain = {
        ...baseProps,
        chainId: 137, // Polygon
      };

      render(<TransactionDetails {...propsWithDifferentChain} />);

      expect(screen.getByText('Transaction Details')).toBeInTheDocument();
    });
  });

  describe('handles technical details generation', () => {
    it('generates technical details for sell mode', () => {
      const propsWithError = {
        ...baseProps,
        status: 'Transaction Failed' as const,
        errorDetails: 'Test error message',
      };

      render(<TransactionDetails {...propsWithError} />);

      const technicalDetails = screen.getByTestId('technical-details');
      expect(technicalDetails).toBeInTheDocument();

      const detailsText = technicalDetails.textContent;
      expect(detailsText).toContain('"transactionType": "SELL"');
      expect(detailsText).toContain('"hashType": "userOpHash"');
      expect(detailsText).toContain('"status": "Transaction Failed"');
    });

    it('generates technical details for buy mode', () => {
      const buyPropsWithError = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
        status: 'Transaction Failed' as const,
        errorDetails: 'Test error message',
      };

      render(<TransactionDetails {...buyPropsWithError} />);

      const technicalDetails = screen.getByTestId('technical-details');
      expect(technicalDetails).toBeInTheDocument();

      const detailsText = technicalDetails.textContent;
      expect(detailsText).toContain('"transactionType": "BUY"');
      expect(detailsText).toContain('"hashType": "bidHash"');
      expect(detailsText).toContain('"status": "Transaction Failed"');
    });

    it('sanitizes API keys in error details', () => {
      const propsWithApiKey = {
        ...baseProps,
        status: 'Transaction Failed' as const,
        errorDetails:
          'error inside sendUserOpToBundler: {"details":"Unexpected behaviour","metaMessages":["Status: 500","URL: https://rpc.etherspot.io/v2/8453?api-key=test-api-key-123","Request body: {...}"]}',
      };

      render(<TransactionDetails {...propsWithApiKey} />);

      const technicalDetails = screen.getByTestId('technical-details');
      const detailsText = technicalDetails.textContent;
      expect(detailsText).toContain('api-key=***REDACTED***');
      expect(detailsText).not.toContain('test-api-key-123');
    });
  });

  describe('handles progress step visibility', () => {
    it('hides progress steps when transaction is complete', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Complete" />
      );

      expect(
        screen.queryByTestId('progress-step-Submitted')
      ).not.toBeInTheDocument();
      expect(
        screen.queryByTestId('progress-step-Pending')
      ).not.toBeInTheDocument();
      expect(
        screen.queryByTestId('progress-step-Completed')
      ).not.toBeInTheDocument();
    });

    it('shows progress steps when transaction is not complete', () => {
      render(
        <TransactionDetails {...baseProps} status="Transaction Pending" />
      );

      expect(screen.getByTestId('progress-step-Submitted')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Pending')).toBeInTheDocument();
      expect(screen.getByTestId('progress-step-Completed')).toBeInTheDocument();
    });
  });
});
