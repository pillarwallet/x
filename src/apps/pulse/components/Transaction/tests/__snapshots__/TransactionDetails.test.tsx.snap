// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TransactionDetails /> > Rendering > renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col gap-4 h-full w-full"
  data-testid="pulse-transaction-details-container"
>
  <div
    className="flex justify-between items-center"
  >
    <p
      className="text-xl text-white font-normal"
      data-testid="pulse-transaction-details-title"
    >
      Transaction Details
    </p>
    <div
      className="justify-center items-center bg-[#121116] rounded-[10px] p-[2px_2px_4px_2px] flex w-10 h-10 ml-3"
      data-testid="pulse-transaction-details-close-button"
    >
      <button
        aria-label="Close"
        className="flex items-center justify-center w-[36px] h-[34px] bg-[#1E1D24] rounded-[8px]"
        data-testid="close-button"
        onClick={[MockFunction spy]}
        type="button"
      >
        <svg
          className="w-5 h-5 text-white/50"
          fill="none"
          stroke="currentColor"
          strokeWidth="2.5"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="flex flex-col h-full w-full rounded-[10px] border border-dashed border-[#25232D] p-6 items-center justify-center"
    data-testid="pulse-transaction-details-summary-sell-Test Token"
  >
    <div
      className="text-5xl text-white font-medium"
      data-testid="pulse-transaction-details-amount-usd"
    >
      <span
        className="text-white/30"
      >
        $
      </span>
      50
    </div>
    <div
      className="flex gap-1.5 items-center"
      data-testid="pulse-transaction-details-token-info"
    >
      <img
        alt="TEST"
        className="w-4 h-4 rounded"
        data-testid="pulse-transaction-details-token-logo-Test Token"
        src="test-logo.png"
      />
      <p
        className="text-white font-normal text-[13px]"
        data-testid="pulse-transaction-details-token-amount"
      >
        100
         
        <span
          className="text-white/30"
          data-testid="pulse-transaction-details-token-symbol"
        >
          TEST
        </span>
      </p>
    </div>
  </div>
  <div
    className="flex h-full w-full rounded-[10px] border border-dashed border-[#25232D] p-3"
    data-testid="pulse-transaction-details-progress-container"
  >
    <div
      className="flex flex-col items-center justify-center w-full"
    >
      <div
        className="flex flex-col w-full"
        data-testid="pulse-transaction-details-progress-steps"
      >
        <div
          data-testid="progress-step-Submitted"
        >
          <span
            data-testid="step-Submitted-status"
          >
            completed
          </span>
          <span
            data-testid="step-Submitted-label"
          >
            Submitted
          </span>
          <div
            data-testid="step-Submitted-timestamp"
          >
            Formatted Submitted timestamp: 2023-01-01T00:00:00.000Z
          </div>
        </div>
        <div
          data-testid="progress-step-Pending"
        >
          <span
            data-testid="step-Pending-status"
          >
            pending
          </span>
          <span
            data-testid="step-Pending-label"
          >
            Pending
          </span>
        </div>
        <div
          data-testid="progress-step-Completed"
        >
          <span
            data-testid="step-Completed-status"
          >
            pending
          </span>
          <span
            data-testid="step-Completed-label"
          >
            Completed
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="flex flex-col h-full w-full rounded-[10px] border border-dashed border-[#25232D] p-3"
    data-testid="pulse-transaction-details-info-container"
  >
    <div
      data-testid="transaction-info"
    >
      <span
        data-testid="info-status"
      >
        Starting Transaction
      </span>
      <span
        data-testid="info-userOpHash"
      >
        0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef
      </span>
      <span
        data-testid="info-txHash"
      />
      <span
        data-testid="info-chainId"
      >
        1
      </span>
      <span
        data-testid="info-gasFee"
      >
        0.001
      </span>
    </div>
  </div>
  <div
    className="w-full rounded-[10px] bg-[#121116] p-[2px_2px_6px_2px]"
    data-testid="pulse-transaction-details-done-button-container"
  >
    <button
      className="flex items-center justify-center w-full rounded-[8px] h-[42px] p-[1px_6px_1px_6px] bg-[#8A77FF] text-white font-normal text-[14px]"
      data-testid="pulse-transaction-details-done-button"
      onClick={[MockFunction spy]}
      type="button"
    >
      Done
    </button>
  </div>
</div>
`;
